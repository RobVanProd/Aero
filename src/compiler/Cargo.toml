[package]
name = "compiler"
version = "0.2.0"
edition = "2024"

[lib]
path = "src/lib.rs"

[dependencies]
inkwell = { git = "https://github.com/TheDan64/inkwell", branch = "master" }
thiserror = "1.0"
miette = { version = "7.0", features = ["fancy"] }
ariadne = "0.4"

[[bin]]
name = "aero"
path = "src/main.rs"

[dev-dependencies]
insta = "1.34.0"
compiler = { path = "." }


