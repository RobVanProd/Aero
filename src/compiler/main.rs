// src/compiler/main.rs

fn main() {
    println!("Aero compiler prototype");
}


