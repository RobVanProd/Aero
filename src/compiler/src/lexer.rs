pub fn tokenize(source: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut chars = source.chars().peekable();

    while let Some(&c) = chars.peek() {
        match c {
            // Whitespace
            ' ' | '\t' | '\n' | '\r' => {
                chars.next();
            }
            // Operators and delimiters
            '+' => { tokens.push("+".to_string()); chars.next(); }
            '-' => { tokens.push("-".to_string()); chars.next(); }
            '*' => { tokens.push("*".to_string()); chars.next(); }
            '/' => { tokens.push("/".to_string()); chars.next(); }
            '%' => { tokens.push("%".to_string()); chars.next(); }
            '=' => { tokens.push("=".to_string()); chars.next(); }
            ';' => { tokens.push(";".to_string()); chars.next(); }
            '{' => { tokens.push("{".to_string()); chars.next(); }
            '}' => { tokens.push("}".to_string()); chars.next(); }
            '(' => { tokens.push("(".to_string()); chars.next(); }
            ')' => { tokens.push(")".to_string()); chars.next(); }
            // Float literals starting with decimal point (e.g., .5)
            '.' => {
                chars.next(); // consume the '.'
                if let Some(&d) = chars.peek() {
                    if d.is_ascii_digit() {
                        let mut num_str = String::from("0.");
                        while let Some(&digit) = chars.peek() {
                            if digit.is_ascii_digit() {
                                num_str.push(chars.next().unwrap());
                            } else {
                                break;
                            }
                        }
                        // Handle scientific notation (e.g., .5e3)
                        if let Some(&e_char) = chars.peek() {
                            if e_char == 'e' || e_char == 'E' {
                                num_str.push(chars.next().unwrap());
                                if let Some(&sign) = chars.peek() {
                                    if sign == '+' || sign == '-' {
                                        num_str.push(chars.next().unwrap());
                                    }
                                }
                                while let Some(&digit) = chars.peek() {
                                    if digit.is_ascii_digit() {
                                        num_str.push(chars.next().unwrap());
                                    } else {
                                        break;
                                    }
                                }
                            }
                        }
                        tokens.push(format!("float_literal:{}", num_str));
                    } else {
                        // Just a dot, not a float literal
                        tokens.push(".".to_string());
                    }
                } else {
                    tokens.push(".".to_string());
                }
            }
            // Integer and Float Literals
            '0'..='9' => {
                let mut num_str = String::new();
                let mut has_dot = false;
                let mut has_exponent = false;
                
                // Collect digits and decimal point
                while let Some(&d) = chars.peek() {
                    if d.is_ascii_digit() {
                        num_str.push(chars.next().unwrap());
                    } else if d == '.' && !has_dot && !has_exponent {
                        has_dot = true;
                        num_str.push(chars.next().unwrap());
                    } else if (d == 'e' || d == 'E') && !has_exponent {
                        has_exponent = true;
                        num_str.push(chars.next().unwrap());
                        // Handle optional sign after exponent
                        if let Some(&sign) = chars.peek() {
                            if sign == '+' || sign == '-' {
                                num_str.push(chars.next().unwrap());
                            }
                        }
                    } else {
                        break;
                    }
                }
                
                if has_dot || has_exponent {
                    tokens.push(format!("float_literal:{}", num_str));
                } else {
                    tokens.push(format!("integer_literal:{}", num_str));
                }
            }
            // Identifiers and Keywords
            'a'..='z' | 'A'..='Z' | '_' => {
                let mut ident_str = String::new();
                while let Some(&d) = chars.peek() {
                    if d.is_ascii_alphanumeric() || d == '_' {
                        ident_str.push(chars.next().unwrap());
                    } else {
                        break;
                    }
                }
                match ident_str.as_str() {
                    "let" => tokens.push("keyword:let".to_string()),
                    "fn" => tokens.push("keyword:fn".to_string()),
                    "return" => tokens.push("keyword:return".to_string()),
                    _ => tokens.push(format!("identifier:{}", ident_str)),
                }
            }
            _ => {
                // Handle unexpected characters or errors
                eprintln!("Unexpected character: {}", c);
                chars.next();
            }
        }
    }
    tokens
}


