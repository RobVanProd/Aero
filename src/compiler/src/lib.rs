mod ast;
mod lexer;
mod parser;
mod semantic_analyzer;
mod ir;
mod ir_generator;
mod code_generator;
mod types;
