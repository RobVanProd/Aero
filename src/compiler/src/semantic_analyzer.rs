use std::collections::HashMap;
use crate::ast::{AstNode, Expression, Statement};
use crate::types::{Ty, infer_binary_type};

#[derive(Debug, Clone)]
pub struct VarInfo {
    pub ptr_name: String,
    pub ty: Ty, // Changed from String to Ty
    pub initialized: bool,
}

pub struct SemanticAnalyzer {
    symbol_table: HashMap<String, VarInfo>, // Stores variable_name -> VarInfo
    next_ptr_id: u32,
}

impl SemanticAnalyzer {
    pub fn new() -> Self {
        SemanticAnalyzer { 
            symbol_table: HashMap::new(),
            next_ptr_id: 0,
        }
    }

    fn fresh_ptr_name(&mut self) -> String {
        let ptr_name = format!("ptr{}", self.next_ptr_id);
        self.next_ptr_id += 1;
        ptr_name
    }

    pub fn analyze(&mut self, mut ast: Vec<AstNode>) -> Result<(String, Vec<AstNode>), String> {
        // First pass: type inference and validation
        for node in &mut ast {
            match node {
                AstNode::Statement(stmt) => {
                    match stmt {
                        Statement::Let { name, value } => {
                            if self.symbol_table.contains_key(name) {
                                return Err(format!("Error: Variable `{}` already declared.", name));
                            }
                            
                            // Infer and validate the expression type
                            let inferred_type = self.infer_and_validate_expression(value)?;
                            let ptr_name = self.fresh_ptr_name();
                            let var_info = VarInfo {
                                ptr_name: ptr_name.clone(),
                                ty: inferred_type.clone(),
                                initialized: true, // Let statements always initialize
                            };
                            self.symbol_table.insert(name.clone(), var_info);
                            println!("Declared variable: {} with type {} (ptr: {})", name, inferred_type.to_string(), ptr_name);
                        }
                        Statement::Return(expr) => {
                            // Check if the expression uses any variables and validate they are initialized
                            self.check_expression_initialization(expr)?;
                            self.infer_and_validate_expression(expr)?;
                            println!("Return statement with expression of type inferred.");
                        }
                    }
                }
                AstNode::Expression(_) => {
                    // Top-level expressions are not handled yet
                    return Err("Error: Top-level expressions not supported yet.".to_string());
                }
            }
        }
        Ok(("Semantic analysis completed successfully. No borrow checker violations detected.".to_string(), ast))
    }

    fn check_expression_initialization(&self, expr: &Expression) -> Result<(), String> {
        match expr {
            Expression::Number(_) | Expression::Float(_) => Ok(()),
            Expression::Identifier(name) => {
                if let Some(var_info) = self.symbol_table.get(name) {
                    if !var_info.initialized {
                        Err(format!("Error: Use of uninitialized variable `{}`.", name))
                    } else {
                        Ok(())
                    }
                } else {
                    Err(format!("Error: Use of undeclared variable `{}`.", name))
                }
            }
            Expression::Binary { lhs, rhs, .. } => {
                self.check_expression_initialization(lhs)?;
                self.check_expression_initialization(rhs)?;
                Ok(())
            }
        }
    }

    fn infer_and_validate_expression(&self, expr: &mut Expression) -> Result<Ty, String> {
        match expr {
            Expression::Number(_) => Ok(Ty::Int),
            Expression::Float(_) => Ok(Ty::Float),
            Expression::Identifier(name) => {
                if let Some(var_info) = self.symbol_table.get(name) {
                    if !var_info.initialized {
                        Err(format!("Error: Use of uninitialized variable `{}`.", name))
                    } else {
                        Ok(var_info.ty.clone())
                    }
                } else {
                    Err(format!("Error: Use of undeclared variable `{}`.", name))
                }
            }
            Expression::Binary { op, lhs, rhs, ty } => {
                let lhs_type = self.infer_and_validate_expression(lhs)?;
                let rhs_type = self.infer_and_validate_expression(rhs)?;

                // Infer the result type using type promotion rules
                let result_type = infer_binary_type(op, &lhs_type, &rhs_type)?;
                
                // Store the result type in the AST node
                *ty = Some(result_type.clone());
                
                Ok(result_type)
            }
        }
    }

    pub fn get_var_info(&self, name: &str) -> Option<&VarInfo> {
        self.symbol_table.get(name)
    }
}


