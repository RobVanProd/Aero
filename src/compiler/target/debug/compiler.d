/home/ubuntu/aero_project/src/compiler/target/debug/compiler: /home/ubuntu/aero_project/src/compiler/main.rs
