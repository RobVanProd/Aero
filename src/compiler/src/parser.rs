use crate::ast::{AstNode, Expression, Statement};
use crate::lexer;
use crate::semantic_analyzer;
use crate::ir_generator;
use crate::code_generator;

pub fn parse(tokens: Vec<String>) -> Vec<AstNode> {
    let mut ast_nodes = Vec::new();
    let mut i = 0;

    while i < tokens.len() {
        match tokens[i].as_str() {
            "keyword:let" => {
                // Expecting 'let identifier = expression;'
                i += 1; // consume 'let'
                let name_token = tokens[i].clone(); // identifier name token
                let name = name_token.split(':').nth(1).unwrap().to_string(); // Extract just the name
                i += 1;
                if tokens[i] != "=" {
                    eprintln!("Expected = after identifier in let statement");
                    break; // Use break instead of return to continue parsing other statements
                }
                i += 1; // consume '='

                let mut expr_tokens = Vec::new();
                while i < tokens.len() && tokens[i] != ";" {
                    expr_tokens.push(tokens[i].clone());
                    i += 1;
                }

                let expr = parse_expression(expr_tokens);

                if i >= tokens.len() || tokens[i] != ";" {
                    eprintln!("Expected ; after expression in let statement");
                    break; // Use break instead of return
                }
                i += 1; // consume ';'

                ast_nodes.push(AstNode::Statement(Statement::Let { name, value: expr }));
            }
            _ => {
                eprintln!("Unexpected token at top level: {}", tokens[i]);
                i += 1;
            }
        }
    }

    ast_nodes
}

fn parse_expression(expr_tokens: Vec<String>) -> Expression {
    // Simple expression parsing for now (only numbers, identifiers, and binary ops)
    // This needs to be replaced with a proper Pratt parser or similar for operator precedence
    let mut i = 0;
    let mut lhs = match expr_tokens[i].split(':').next().unwrap() {
        "integer_literal" => Expression::Number(expr_tokens[i].split(':').nth(1).unwrap().parse().unwrap()),
        "float_literal" => Expression::Float(expr_tokens[i].split(':').nth(1).unwrap().parse().unwrap()),
        "identifier" => Expression::Identifier(expr_tokens[i].split(':').nth(1).unwrap().to_string()),
        _ => panic!("Unsupported expression token: {}", expr_tokens[i]),
    };
    i += 1;

    while i < expr_tokens.len() {
        let op = expr_tokens[i].clone();
        i += 1;
        let rhs = match expr_tokens[i].split(':').next().unwrap() {
            "integer_literal" => Expression::Number(expr_tokens[i].split(':').nth(1).unwrap().parse().unwrap()),
            "float_literal" => Expression::Float(expr_tokens[i].split(':').nth(1).unwrap().parse().unwrap()),
            "identifier" => Expression::Identifier(expr_tokens[i].split(':').nth(1).unwrap().to_string()),
            _ => panic!("Unsupported expression token: {}", expr_tokens[i]),
        };
        i += 1;
        lhs = Expression::Binary { op, lhs: Box::new(lhs), rhs: Box::new(rhs) };
    }
    lhs
}


