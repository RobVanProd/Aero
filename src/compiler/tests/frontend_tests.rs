use compiler::lexer;
use compiler::parser;
use compiler::semantic_analyzer::{SemanticAnalyzer};
use insta::assert_debug_snapshot;

#[test]
fn test_lexer_simple_let() {
    let source = "let x = 10;";
    let tokens = lexer::tokenize(source);
    assert_debug_snapshot!(tokens);
}

#[test]
fn test_lexer_binary_expression() {
    let source = "let y = x + 5.0;";
    let tokens = lexer::tokenize(source);
    assert_debug_snapshot!(tokens);
}

#[test]
fn test_parser_simple_let() {
    let source = "let x = 10;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    assert_debug_snapshot!(ast);
}

#[test]
fn test_parser_binary_expression() {
    let source = "let y = x + 5.0;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    assert_debug_snapshot!(ast);
}

#[test]
fn test_semantic_simple_let() {
    let source = "let x = 10;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    let mut analyzer = SemanticAnalyzer::new();
    let result = analyzer.analyze(ast);
    assert_debug_snapshot!(result);
}

#[test]
fn test_semantic_binary_expression() {
    let source = "let x = 10; let y = x + 5.0;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    let mut analyzer = SemanticAnalyzer::new();
    let result = analyzer.analyze(ast);
    assert_debug_snapshot!(result);
}

#[test]
fn test_semantic_redeclaration() {
    let source = "let x = 10; let x = 20;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    let mut analyzer = SemanticAnalyzer::new();
    let result = analyzer.analyze(ast);
    assert_debug_snapshot!(result);
}

#[test]
fn test_semantic_undeclared_variable() {
    let source = "let y = x + 10;";
    let tokens = lexer::tokenize(source);
    let ast = parser::parse(tokens);
    let mut analyzer = SemanticAnalyzer::new();
    let result = analyzer.analyze(ast);
    assert_debug_snapshot!(result);
}


