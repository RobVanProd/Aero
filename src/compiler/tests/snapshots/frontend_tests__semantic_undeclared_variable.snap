---
source: tests/frontend_tests.rs
expression: result
---
Err(
    "Error: Use of undeclared variable `x`.",
)
